kernel/src/grant.rs

```diff
@@ -240,75 +449,202 @@ impl<'a> GrantUpcallTable<'a> {
-pub(crate) fn subscribe(
-    process: &dyn Process,
-    upcall: Upcall,
-) -> Result<Upcall, (Upcall, ErrorCode)> {
-    let grant_num = match process.lookup_grant_from_driver_num(upcall.upcall_id.driver_num) {
-        Ok(grant_num) => grant_num,
-        Err(e) => return Err((upcall, e.into())),
-    };
+impl Default for SavedUpcall {
+    fn default() -> Self {
+        Self {
+            appdata: 0,
+            fn_ptr: None,
+        }
+    }
+}
 
-    // Check if the grant has been allocated, and if not we cannot handle the
-    // subscribe call.
-    if let Some(is_allocated) = process.grant_is_allocated(grant_num) {
-        if !is_allocated {
-            return Err((upcall, ErrorCode::NOMEM));
+/// A minimal representation of a read-only allow from app, used for
+/// storing a read-only allow in a process' kernel managed grant space
+/// without wasting memory duplicating information such as process ID.
+#[repr(C)]
+struct SavedAllowRo {
+    ptr: *const u8,
+    len: usize,
+}
+
+impl Default for SavedAllowRo {
+    fn default() -> Self {
+        Self {
+            ptr: core::ptr::null(),
+            len: 0,
         }
-    } else {
-        // Process is no longer active, this case will never happen.
-        return Err((upcall, ErrorCode::FAIL));
     }
+}
 
-    // Return early if no grant.
-    let grant_ptr = match process.enter_grant(grant_num) {
-        Ok(grant_ptr) => grant_ptr,
-        Err(_) => return Err((upcall, ErrorCode::NOMEM)),
+/// A minimal representation of a read-write allow from app, used for
+/// storing a read-write allow in a process' kernel managed grant space
+/// without wasting memory duplicating information such as process ID.
+#[repr(C)]
+struct SavedAllowRw {
+    ptr: *mut u8,
+    len: usize,
+}
+
+impl Default for SavedAllowRw {
+    fn default() -> Self {
+        Self {
+            ptr: core::ptr::null_mut(),
+            len: 0,
+        }
+    }
+}
+
+/// Write the default value of T to every element of the array.
+///
+/// # Safety
+///
+/// The pointer must be well aligned and point to allocated memory that is writable
+/// for `size_of::<T> * num` bytes. The memory does not need to be initialized yet. If
+/// it already does contain initialized memory, then those contents will be overwritten
+/// without being `Drop`ed first.
+unsafe fn write_default_array<T: Default>(base: *mut T, num: usize) {
+    for i in 0..num {
+        base.add(i).write(T::default());
+    }
+}
+
+/// Lifetime of guard represents the lifetime a grant is help "open". On Drop, we leave grant
+struct GrantEnterLifetimeGuard<'a>(&'a dyn Process, usize);
+
+impl Drop for GrantEnterLifetimeGuard<'_> {
+    fn drop(&mut self) {
+        self.0.leave_grant(self.1);
+    }
+}
+
+/// Enters the grant for the specified process. Caller MUST leave_grant with the first element
+/// of the Ok return tuple, as it is the grant_num. The second element of the Ok result is
```

Comment:
I debated this because it would be logically incorrect to enter without a leave, but ultimately it doesn't create any unsoundness issues. It would poison the grant and would prevent accessing the grant in the future, but nothing would be unsound in the rust sense.